### PR TITLE
stat module: fix API inconsistency, add basic test

### PR DESCRIPTION
**Pull request by titodalcanton**
@MPillas and I noticed that one of the simple coinc ranking stats are not usable with the multi-ifo workflow. When we tried using the `phasetd_newsnr_sgveto` stat, we got this error from some of the coinc jobs:
```
 Traceback (most recent call last):
  File "somewhere/pycbc-v1.16.13/bin/pycbc_multiifo_coinc_findtrigs", line 321, in 
    time_addition=args.coinc_threshold)
  File "somewhere/pycbc-v1.16.13/lib/python2.7/site-packages/pycbc/events/stat.py", line 151, in coinc_multiifo
    return sum(sngl[1] ** 2. for sngl in s) ** 0.5
  File "somewhere/pycbc-v1.16.13/lib/python2.7/site-packages/pycbc/events/stat.py", line 151, in 
    return sum(sngl[1] ** 2. for sngl in s) ** 0.5
TypeError: ufunc 'square' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

I *think* I spotted where this is coming from. Basically it seems that some of the stats expect the single-detector triggers to come in this format:
```
[(ifo, iterable of single-detector ranks),
 (ifo, iterable of single-detector ranks)]
```
while others expect this:
```
[(ifo, {'snglstat': iterable of single-detector ranks}),
 (ifo, {'snglstat': iterable of single-detector ranks})]
```
The error was observed with PyCBC 1.16.13, prior to the stat reorganization, but it seems the issue is still present on master.

This PR changes all stats to take input in the latter format. Since PyCBC Live was actually using the former format, I also changed it to be consistent with the latter. Note that I only changed the `rank_stat_coinc()` methods, and I am not sure if other places need the same change.

In addition, given that we should start working on #3584 soon, I added a basic unit test file for the stat module to hopefully make that easier. It is currently failing due to the various unimplemented stats, and the need for some stats to have their files passed to them, so those are explicitly excluded from the test for now.

Edit: made explanation hopefully clearer.

### CODE REVIEW

**Comment by titodalcanton** [comment]
Hmm ok, apparently PyCBC Live does not like the "fix". Putting on hold until someone has a chance to comment.

**Comment by titodalcanton** [comment]
Looks like PyCBC Live needed just a trivial fix - removing the on-hold label.

**Comment by titodalcanton** [comment]
Also tagging @GarethDaviesGW and @tdent.

File: pycbc/events/stat.py [hunk 1]
```diff
@@ -240,7 +240,7 @@ def rank_stat_coinc(self, sngls_list, slide, step, to_shift,
         numpy.ndarray
             Array of coincident ranking statistic values
         """
-        cstat = sum(sngl[1] ** 2. for sngl in sngls_list) ** 0.5
```
Comment (spxiwh) [inline pycbc/events/stat.py:243]: I think this change is going to break this statistic in the standard workflow, where this statistic did work before and did not have a ['snglstat'] element.

Here, I've not changed the previous behaviour (although one might argue a change might be needed), where you call through to the `single` method of the Stat class to get whatever that returns (which can be computed without knowing anything about the other ifo(s), and then feeding that through to `rank_coinc`. For the QuadratureSum method you have not changed the `single` method, but have changed  the `rank_coinc` method, this breaks this statistic.

If possible, PyCBC Live should be doing the same and not manually constructing the single_info as you do on line 1046 of coinc.py but instead using the `single` method of `self.stat_calculator` and then passing that along .... Maybe there's some reason why it needs to do this though??

**Review by spxiwh — CHANGES_REQUESTED** [review]
I don't think this is the right fix to the problem (see comments in line). This will definitely break the QuadratureSum statistic.

A possible solution is to enforce a common format for the output of `single` method, but we did try that initially, and it didn't work so we had to change it. Maybe the dictionary format is better though, and could be enforced for all statistics. (I emphasize though that this behaviour is unchanged from before ...)

**Comment by spxiwh** [comment]
@titodalcanton I read the code changes and didn't really read the comments. Indeed the `single` method (before and after refactoring) did return inconsistent output. But as this would only be used *within* each stat class, it was not a problem. The user should not have had to care as you should just let the stat class choose.

I think the prime problem here is that PyCBC Live is trying to guess this format itself. It shouldn't be doing any form of stat calculations (although I'm probably missing a reason why it needs to).

That said, I think the output from `single` perhaps could be unified. But I think what's here now will break the QuadratureSum class.

**Pull request closed without merging**